### PR TITLE
Ensure TokenCredentials are serializable to JSON

### PR DESCRIPTION
**Pull request by hsartoris-bard**
See also #2670.

Currently, the second test will fail unless additional typing information is introduced. Happy to tweak as needed.

### CODE REVIEW

**Comment by leleuj** [comment]
Good. What's your fix proposal?

**Comment by hsartoris-bard** [comment]
https://github.com/pac4j/pac4j/pull/2692/commits/c2cff70ad78995ffae92b448875529f52a644361 provides a fix for the first test, which otherwise fails.

https://github.com/pac4j/pac4j/pull/2692/commits/16d55922b2a424b9703152100e835024ca857fbd addresses the latter, by explicitly annotating `UserProfile` with `@JsonTypeInfo`. I also added `@EqualsAndHashCode` on `BasicUserProfile`, because otherwise equality checking on `TokenCredentials` doesn't work, but this is optional - I can either check that the tokens match between the original and the deserialized credentials, or set `callSuper` to false on TokenCredentials. My sense is that, semantically, it doesn't make sense to say that two TokenCredentials objects are equal simply if their tokens are equal, even if they contain differing UserProfiles.

**Comment by leleuj** [comment]
Can you check why the build fails: https://github.com/pac4j/pac4j/actions/runs/6470734206/job/17595797077?pr=2692?

**Comment by hsartoris-bard** [comment]
Ah yes, needed to also override `equals` on CasProxyProfile

**Comment by leleuj** [comment]
The build still fails: `Error:  Medium: org.pac4j.kerberos.profile.KerberosProfile doesn't override org.pac4j.core.profile.BasicUserProfile.equals(Object) [org.pac4j.kerberos.profile.KerberosProfile] At KerberosProfile.java:[line 1] EQ_DOESNT_OVERRIDE_EQUALS`

File: pac4j-cas/src/main/java/org/pac4j/cas/profile/CasProxyProfile.java [hunk 1]
```diff
@@ -16,6 +17,7 @@
  * @author Jerome Leleu
  * @since 1.4.0
  */
+@EqualsAndHashCode
```
Comment (mmoayyed) [inline pac4j-cas/src/main/java/org/pac4j/cas/profile/CasProxyProfile.java:20]: Should this not be `@EqualsAndHashCode(callSuper = true)`?

File: pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java [hunk 2]
```diff
@@ -12,6 +14,7 @@
  * @author Jerome Leleu
  * @since 4.0.0
  */
+@JsonTypeInfo(use=JsonTypeInfo.Id.CLASS, include=JsonTypeInfo.As.PROPERTY, property="class")
```
Comment (mmoayyed) [inline pac4j-core/src/main/java/org/pac4j/core/profile/UserProfile.java:17]:
```suggestion
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "class")
```

**Review by mmoayyed — CHANGES_REQUESTED** [review]

**Comment by hsartoris-bard** [comment]
good points; changes made. I was able to run the tests locally by excluding the SAML test suite, which appears to expect some service to be running that I do not have.

**Comment by leleuj** [comment]
Can you sync with master? I have excluded two failing SAML tests.

**Comment by hsartoris-bard** [comment]
Done; tests all pass for me now.

**Pull request merged**